---
title: "`phydms` Tutorial"
output:
  html_document:
#    toc: true
#    toc_depth: 2
#    toc_float: true
    highlight: pygments
    theme: spacelab
    code_folding: hide
---

In this tutorial you will learn the basics of `phydms`. We will walk through examples of analyses you may want to run to compare your deep mutational scanning data to natural sequence evolution. For more details on any of the steps, please see the [phydms documentation](http://jbloomlab.github.io/phydms/).     

`phydms` was developed at the [Bloom Lab](http://research.fhcrc.org/bloom/en.html) ([full list of contributors](https://github.com/jbloomlab/phydms/graphs/contributors))

#Example Analyses{.tabset .tabset-fade}  
The examples used here come from 
[Bloom, Biology Direct, 12:1](http://biologydirect.biomedcentral.com/articles/10.1186/s13062-016-0172-z)

##Comparing your `DMS` data to natural sequence evolution  
After you perform your deep mutational scanning experiment, you may want to know how well the experimental measurements you made in the lab describe the natural sequence evolution of your protein in nature. Using `phydms` you can compare the phylogenetic substitution model [*ExpCM*](https://academic.oup.com/mbe/article/31/10/2753/1015257/An-Experimentally-Informed-Evolutionary-Model), which takes into account the site-specific amino-acid preferences from your deep mutational scanning experiment, to traditional, non-site specific models from the [*YNGKP*](http://www.genetics.org/content/155/1/431) family. [`phydms_comprehensive`](http://jbloomlab.github.io/phydms/phydms_comprehensive_prog.html) will run `phydms` in several different modes to generate results for the appropriate comparisons.   

For the standard `phydms_comprehensive` analysis, you will need [amino-acid preferences](#prefs) from your deep mutational scanning experiment and a codon-level sequence [alignment](#seqs) of your gene. `phydms_comprehensive` will then   

• infer a tree using [`RAxML`](#http://sco.h-its.org/exelixis/software.html)  
• run `phydms` with the *YNGKP_M0*, the *YNKGP_M5*, the *ExpCM*, and a control *ExpCM* run with averaged preferences  
• summarize the results  

See the [full documentation](http://jbloomlab.github.io/phydms/index.html) for specifics on these models.  

We are going to walk through an example of `phydms_comprehensive` and evaluation of how well the $\beta$-lactamase deep mutational scanning data from [Stiffler *et al*, 2015](http://www.sciencedirect.com/science/article/pii/S0092867415000781) describes natural sequence evolution. 

###`phydms_comprehensive` command-line usage
Here is the full list of requirements and options for `phydms_comprehensive`. Below is a discussion of the [input files](#inputFiles), running the [`phydms_comprehensive` command](#phydms_comprehensiveCommand), and [interpretation of the results](#interpretation).   
```{r phydms_comprehensive help, engine='bash', comment=NA, error=TRUE}
phydms_comprehensive -h
```   

###Input Files{#inputFiles}
####Amino-acid preferences{#prefs}
Often data from deep mutational scanning experiments is reported as the enrichment of a given amino-acid compared to the wild-type amino-acid. 
Here is a snippet of the log enrichment scores for $\beta$-lactamase from the file [betaLactamase_enrichmentScores.csv](example_data/betaLactamase_enrichmentScores.csv) in `example_data` and the full dataset visualized as a heatmap.
```{r enrichment scores heatmap, comment=NA}
library(ggplot2)
df = read.csv("example_data/betaLactamase_enrichmentScores.csv")
head(df,20)
p <- ggplot(df, aes(Site, AminoAcid)) + geom_tile(aes(fill = Trial_1_AmpConc_2500),colour = "white") + scale_fill_gradient(low = "white",high = "steelblue")
p
```     

`phydms` uses *amino-acid preferences* rather than *enrichment scores*. We can transform these enrichment scores to amino-acid preferences by normalizing the values for each site. Below is the first few sites and amino-acids of the resulting [preferences](example_data/betaLactamase_prefs.csv). Notice that the numbering has changed from above. In `phydms` a site is numbered in relation to the first site in the preferences rather than to the start codon.   

```{r enrichment scores to preferences, engine="python", comment=NA}
import pandas as pd
df = pd.read_csv("example_data/betaLactamase_enrichmentScores.csv")
minenrichment = 1.0e-4 # minimum allowed enrichment
df["preference"] = [max(minenrichment, 10**df["Trial_1_AmpConc_2500"][x] + 10**df["Trial_1_AmpConc_2500"][x]) for x in range(len(df))]
df = df.pivot(index = "Site", columns = "AminoAcid", values = "preference")
df.fillna(1, inplace = True)
df = df.div(df.sum(axis=1), axis=0)
df.insert(0, "site", range(1,len(df)+1))
df.to_csv("example_data/betaLactamase_prefs.csv", index = False)
print df.iloc[:,:9].head(20).to_string(index = False)
```   

####Sequences{#seqs}
We will use the [alignment](#example_data/betaLactamase_alignment.fasta) of $\beta$-lactamase sequences found in the `example_data` directory. It is important to note there is exactly the same number of preferences as sites in the alignment. 
```{r alignment, engine="python", comment=NA, echo = FALSE}   
import pandas as pd
fileName = "example_data/betaLactamase_alignment.fasta"
seqTotal = 0
seqLength = []
with open(fileName) as f:
  for line in f:
    if line.startswith(">"):
      seqTotal += 1
    else:
      seqLength.append(len(line.strip()))
assert len(set(seqLength)) == 1
print "Sequences read from: example_data/betaLactamase_alignment.fasta"
print "There are %s sequences"%seqTotal
print "Each sequence is %s amino-acids long"%str(seqLength[0]/3)
print "Preferences read from: example_data/betaLactamase_prefs.csv"
print "There are preferences measured for %s sites"%len(pd.read_csv("example_data/betaLactamase_prefs.csv"))
```  

####`phydms_comprehensive`{#phydms_comprehensiveCommand}
We can now run `phydms_comprehensive` by specifying our output prefix (in this case a directory), our [preferences](#prefs), and our [alignment](#seqs). The output below is the `phydms_comprehensive` run log.  
```
phydms_comprehensive betaLactamase/ example_data/betaLactamase_alignment.fasta example_data/betaLactamase_prefs.csv
```
```{r phydms_comprehensive log, engine='bash', comment=NA, error=TRUE}
cat betaLactamase/log.log
```   
<!-- ```{r phydms_comprehensive run, engine='bash', comment=NA, error=TRUE} -->
<!-- phydms_comprehensive betaLactamase/ example_data/betaLactamase_alignment.fasta example_data/betaLactamase_prefs.csv --raxml -->
<!-- ```    -->

###Output files
`phydms_comprehensive` produces both the standard `phydms` output files for each model and a summary file. 
```{r phydms_comprehensive output files, engine='bash', comment=NA, error=TRUE}
ls betaLactamase/*
```   

###Interpretation of `phydms_comprehensive` results{#interpretation}
To evaluate how well the $\beta$-lactamase deep mutational scanning data describe natural sequence evolution we can look at the summary file, [`modelcomparison.md`](betaLactamase/modelcomparison.md).   


| Model                              | deltaAIC | LogLikelihood | nParams | ParamValues                                   |
|------------------------------------|----------|---------------|---------|-----------------------------------------------|
| ExpCM_betaLactamase_prefs          | 0.00     | -3557.66      | 6       | beta=0.79, kappa=2.61, omega=0.51             |
| YNGKP_M5                           | 350.80   | -3727.06      | 12      | alpha_omega=0.33, beta_omega=0.41, kappa=3.02 |
| averaged_ExpCM_betaLactamase_prefs | 475.92   | -3795.62      | 6       | beta=0.75, kappa=2.54, omega=0.35             |
| YNGKP_M0                           | 507.68   | -3806.50      | 11      | kappa=2.57, omega=0.35                        |


**First, we can see that the *ExpCM* has the lowest log-likelihood of all four of the models.** It significantly outperforms (evaluated by the $\Delta$AIC) the non-site-specific *YNGKP* models. It also outperforms the *ExpCM* control where the preferences are averaged across the sites. **These comparisons are evidence that the deep mutational scanning results do a good job of describing $\beta$-lactamase's evolution in nature.**  

We can also evaluate the amino-acid preferences is by the value of the *ExpCM* stringency parameter, $\beta$. $\beta$ is a way to gauge how well the selection in the lab compares to selection in nature. When $\beta$ is fit to be greater than $1$ it means the selection in lab was not as strong as selection in nature. The converse is true when $\beta$ is less than $1$. In the $\beta$-lactamase example, $\beta$ is fit to be $0.79$. **Since this number is close to $1$, we can conclude that not only do the amino-acid preferences do a good job of describing the natural sequence evolution of $\beta$-lactamase but that the selection pressures exerted in lab closely match the selection pressures in nature. **

Using `phydms_comprehensive` we were able to conclude that the amino-acid preferences (normalized enrichment values) from a deep mutational experiment did a good job of describing the natural sequence evolution of $\beta$-lactamase. 

Please see the [`phydms_comprehensive' full documentation](http://jbloomlab.github.io/phydms/phydms_comprehensive_prog.html) if you would like to learn more about the `phydms_comprehensive` program and its other options. 
 
##Comparing two `DMS` dataset for the same protein
##Detecting Diversifying Selection
