<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>the phydms_testdivpressure program &#8212; phydms 2.0.dev0 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.0.dev0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Experimentally Informed Codon Models" href="ExpCM.html" />
    <link rel="prev" title="the phydms_logoplot program" href="phydms_logoplot.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="ExpCM.html" title="Experimentally Informed Codon Models"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="phydms_logoplot.html" title="the phydms_logoplot program"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">phydms 2.0.dev0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="auxiliary_progs.html" accesskey="U">auxiliary programs</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="the-phydms-testdivpressure-program">
<span id="phydms-testdivpressure-prog"></span><h1><a class="toc-backref" href="#id1">the <code class="docutils literal"><span class="pre">phydms_testdivpressure</span></code> program</a><a class="headerlink" href="#the-phydms-testdivpressure-program" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#the-phydms-testdivpressure-program" id="id1">the <code class="docutils literal"><span class="pre">phydms_testdivpressure</span></code> program</a><ul>
<li><a class="reference internal" href="#overview" id="id2">Overview</a></li>
<li><a class="reference internal" href="#command-line-usage" id="id3">Command-line usage</a></li>
<li><a class="reference internal" href="#output-files" id="id4">Output files</a><ul>
<li><a class="reference internal" href="#log-file" id="id5">Log file</a></li>
<li><a class="reference internal" href="#model-comparison-files" id="id6">Model comparison files</a></li>
<li><a class="reference internal" href="#output-for-individual-models" id="id7">Output for individual models</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id2">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">phydms_testdivpressure</span></code> is a program that simplifies usage of <code class="docutils literal"><span class="pre">phydms</span></code> to compare different models of diversifying pressure.</p>
<p>In its simplest usage, you provide <code class="docutils literal"><span class="pre">phydms_testdivpressure</span></code> with an alignment, a file giving site-specific amino-acid preferences, and a file giving site-specific diversifying pressures.
The program then first uses <code class="docutils literal"><span class="pre">RAxML</span></code> to infer a tree under the <em>GTRCAT</em> model.
Alternatively, you can specify the tree using <code class="docutils literal"><span class="pre">--tree</span></code> flag.
Then, each branch length of the tree is optimized with a control <em>ExpCM</em> without any diversifying pressures.
This optimized tree&#8217;s branch lengths are then scaled with a single parameter under <em>ExpCM</em> for each of the site-specific diversifying pressures.
For a control, <code class="docutils literal"><span class="pre">phydms_testdivpressure</span></code> will also optimize the tree with randomized diversifying pressures.
It creates simple summary files that enable comparison among the models.</p>
<p>See below for information on <a class="reference internal" href="#command-line-usage">Command-line usage</a> and <a class="reference internal" href="#output-files">Output files</a>.</p>
</div>
<div class="section" id="command-line-usage">
<h2><a class="toc-backref" href="#id3">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p>Test different models of diversifying pressure. The &#8216;phydms&#8217; package is written by the Bloom lab (see <a class="reference external" href="https://github.com/jbloomlab/phydms/contributors">https://github.com/jbloomlab/phydms/contributors</a>). Version 2.0.dev0. Full documentation at <a class="reference external" href="http://jbloomlab.github.io/phydms">http://jbloomlab.github.io/phydms</a></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">phydms_testdivpressure</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">tree</span> <span class="n">TREE</span> <span class="o">|</span> <span class="o">--</span><span class="n">raxml</span> <span class="n">RAXML</span><span class="p">]</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">randomizations</span> <span class="n">RANDOMIZATIONS</span><span class="p">]</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">ncpus</span> <span class="n">NCPUS</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                              <span class="n">outprefix</span> <span class="n">alignment</span> <span class="n">prefsfile</span> <span class="n">divpressure</span>
                              <span class="p">[</span><span class="n">divpressure</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Positional arguments:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>outprefix</kbd></td>
<td><p class="first">Prefix for output files.</p>
<p class="last">This prefix can be a directory name (e.g. <code class="docutils literal"><span class="pre">my_directory/</span></code>) if you want to create a new directory. See <a class="reference internal" href="#output-files">Output files</a> for a description of the created files.</p>
</td></tr>
<tr><td class="option-group">
<kbd>alignment</kbd></td>
<td><p class="first">Existing FASTA file with aligned codon sequences.</p>
<p class="last">The alignment must meet the same specifications described in the <code class="docutils literal"><span class="pre">phydms</span></code> documentation for the argument of the same name (see <a class="reference internal" href="phydms_prog.html#phydms-prog"><span class="std std-ref">the phydms program</span></a>).</p>
</td></tr>
<tr><td class="option-group">
<kbd>prefsfile</kbd></td>
<td><p class="first">Existing file with site-specific amino-acid preferences.</p>
<p class="last">Provide the name a file giving site-specific amino-acid preferences. This file should meet the same specifications described in the <code class="docutils literal"><span class="pre">phydms</span></code> documentation for the <em>prefsfile</em> that should accompany an <em>ExpCM</em> (see <a class="reference internal" href="phydms_prog.html#phydms-prog"><span class="std std-ref">the phydms program</span></a>).</p>
</td></tr>
<tr><td class="option-group">
<kbd>divpressure</kbd></td>
<td><p class="first">List of existing files with diversifying pressure at each site</p>
<p class="last">Provide the name of one or more files giving site-specific diversifying pressure values.
The file must specify a diversifying pressure for every site in <code class="docutils literal"><span class="pre">alignment</span></code>, using sequential 1, 2, ... numbering.
The diversifying pressure file can be in either a comma-, tab-, or space-deliminted file with the first column giving the site and the second column giving the diversifying pressure value.
These diversifying pressures will be scaled to fall between 0 and 1. For more information on this scaling procedure see (see <a class="reference internal" href="ExpCM.html#expcm"><span class="std std-ref">Experimentally Informed Codon Models</span></a>)</p>
</td></tr>
</tbody>
</table>
</dd>
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">--tree=False</span></kbd></td>
<td><p class="first">Existing Newick file giving input tree.</p>
<p class="last">If you want to instead fix the tree to some existing topology, use this argument and provide the name of a file giving a valid tree in Newick format. You cannot specify both <code class="docutils literal"><span class="pre">--tree</span></code> and <code class="docutils literal"><span class="pre">--raxml</span></code>.</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--raxml=raxml</span></kbd></td>
<td><p class="first">Path to RAxML (http://sco.h-its.org/exelixis/software.html).</p>
<p class="last">By default, <code class="docutils literal"><span class="pre">phydms_comprehensive</span></code> uses <code class="docutils literal"><span class="pre">RAxML</span></code> to infer a tree topology under the <em>Jukes Cantor</em> model and the command <code class="docutils literal"><span class="pre">raxml</span></code>. Use this argument to specify a path to <code class="docutils literal"><span class="pre">RAxML</span></code> other than <code class="docutils literal"><span class="pre">raxml</span></code>.  The tree inferred by <code class="docutils literal"><span class="pre">RAxML</span></code> can be found in the same directory as the other output files.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--randomizations=0</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Number diversifying pressure randomizations.</p>
<p class="last">This number specifies the number of times the diversifying pressures are randomized. An adequate number of simulations will give an approximate p-value.</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--ncpus=-1</span></kbd></td>
<td>Use this many CPUs; -1 means all available.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-v</span>, <span class="option">--version</span></kbd></td>
<td>show program&#8217;s version number and exit</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="output-files">
<h2><a class="toc-backref" href="#id4">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>Running <code class="docutils literal"><span class="pre">phydms_testdivpressure</span></code> will create the following output files, all with the prefix specified by <code class="docutils literal"><span class="pre">outprefix</span></code>.</p>
<div class="section" id="log-file">
<h3><a class="toc-backref" href="#id5">Log file</a><a class="headerlink" href="#log-file" title="Permalink to this headline">¶</a></h3>
<p>A file with the suffix <code class="docutils literal"><span class="pre">.log</span></code> will be created that summarizes the overall progress of <code class="docutils literal"><span class="pre">phydms_testdivpressure</span></code>. If <code class="docutils literal"><span class="pre">outprefix</span></code> is just a directory name, this file will be called <code class="docutils literal"><span class="pre">log.log</span></code>.</p>
</div>
<div class="section" id="model-comparison-files">
<h3><a class="toc-backref" href="#id6">Model comparison files</a><a class="headerlink" href="#model-comparison-files" title="Permalink to this headline">¶</a></h3>
<p>Two model comparison files will be created, one with the suffix <code class="docutils literal"><span class="pre">modelcomparison.csv</span></code> and one with the suffix <code class="docutils literal"><span class="pre">modelcomparison.md</span></code> to summarize the model fitting.
The first summary file includes all of the parameters of each model.
The second file summarizes the main results for the non-randomized models.</p>
</div>
<div class="section" id="output-for-individual-models">
<h3><a class="toc-backref" href="#id7">Output for individual models</a><a class="headerlink" href="#output-for-individual-models" title="Permalink to this headline">¶</a></h3>
<p>For each individual model, there will also be all of the expected <code class="docutils literal"><span class="pre">phydms</span></code> output files as described in <a class="reference internal" href="phydms_prog.html#phydms-prog"><span class="std std-ref">the phydms program</span></a>. These files will begin with the prefix specified by <code class="docutils literal"><span class="pre">outprefix</span></code>, which will be followed by a description of the model. For instance, if the command is:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">phydms_testdivpressure</span> <span class="n">my_directory</span><span class="o">/</span> <span class="n">alignment</span><span class="o">.</span><span class="n">fasta</span> <span class="n">prefs</span><span class="o">.</span><span class="n">txt</span> <span class="n">divpressure1</span><span class="o">.</span><span class="n">txt</span><span class="p">,</span> <span class="n">divpressure2</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">randomizations</span> <span class="mi">1</span>
</pre></div>
</div>
<p>then we expect output files with the following prefixes:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">my_directory/ExpCM_prefs_*</span></code> : the <em>ExpCM</em> without diversifying pressure.</li>
<li><code class="docutils literal"><span class="pre">my_directory/ExpCM_prefs_divpressure1_True_*</span></code> : the <em>ExpCM</em> using the diversifying pressures in <code class="docutils literal"><span class="pre">divpressure1.txt</span></code>.</li>
<li><code class="docutils literal"><span class="pre">my_directory/ExpCM_prefs_divpressure1_Random_0*</span></code> : the <em>ExpCM</em> using a randomized version of the diversifying pressures in <code class="docutils literal"><span class="pre">divpressure1.txt</span></code>.</li>
<li><code class="docutils literal"><span class="pre">my_directory/ExpCM_prefs_divpressure2_True_*</span></code> : the <em>ExpCM</em> using the diversifying pressures in <code class="docutils literal"><span class="pre">divpressure2.txt</span></code>.</li>
<li><code class="docutils literal"><span class="pre">my_directory/ExpCM_prefs_divpressure2_Random_0*</span></code> : the <em>ExpCM</em> using a randomized version of the diversifying pressures in <code class="docutils literal"><span class="pre">divpressure2.txt</span></code>.</li>
</ul>
</div></blockquote>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">the <code class="docutils literal"><span class="pre">phydms_testdivpressure</span></code> program</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#command-line-usage">Command-line usage</a></li>
<li><a class="reference internal" href="#output-files">Output files</a><ul>
<li><a class="reference internal" href="#log-file">Log file</a></li>
<li><a class="reference internal" href="#model-comparison-files">Model comparison files</a></li>
<li><a class="reference internal" href="#output-for-individual-models">Output for individual models</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="phydms_logoplot.html"
                        title="previous chapter">the <code class="docutils literal"><span class="pre">phydms_logoplot</span></code> program</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="ExpCM.html"
                        title="next chapter"><strong>Exp</strong>erimentally Informed <strong>C</strong>odon <strong>M</strong>odels</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/phydms_testdivpressure_prog.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="ExpCM.html" title="Experimentally Informed Codon Models"
             >next</a> |</li>
        <li class="right" >
          <a href="phydms_logoplot.html" title="the phydms_logoplot program"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">phydms 2.0.dev0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="auxiliary_progs.html" >auxiliary programs</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2015, Jesse D. Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.1.
    </div>
  </body>
</html>